### PR TITLE
fixed run-id diff issue for timescale

### PR DESCRIPTION
**Pull request by Recolada**
Updated logic for Timescale master and worker identification function. 
Added logic to sync run-id between master and worker. only in localrunner mode it works as well.

Tested below

- Ran a test in local runner
- Ran a test in master worker config
- Ran a test by giving run-id in config

### CODE REVIEW

File: locust_plugins/listeners.py [hunk 1]
```diff
@@ -504,9 +519,9 @@ def user_error(self, user_instance, exception, tb, **kwargs):
             self.function(user_instance, exception, tb, **kwargs)
 
 
-def is_worker():
-    return "--worker" in sys.argv
+def is_worker(env: locust.env.Environment):
+    return "--worker" in sys.argv or env.parsed_options.worker
```
[the remark this answers is not in the document]
Comment (Recolada) [reply locust_plugins/listeners.py]: Removed the sys.argv as it is redundant due to env.parsed_options. 
Also handled unused variable warning.

File: locust_plugins/listeners.py [hunk 2]
```diff
@@ -130,22 +133,34 @@ def on_start(self, environment: locust.env.Environment):
...
 
-        if is_worker() or is_master():
+        if is_worker(self.env) or is_master(self.env) or type(environment.runner) == LocalRunner:
             # swarm generates the run id for its master and workers
             if getattr(environment.parsed_options, "run_id", False):
                 self._run_id = parser.parse(environment.parsed_options.run_id)
+            elif hasattr(self, "_run_id"):
+                pass
             else:
                 logging.info(
```
Comment (cyberw) [inline locust_plugins/listeners.py:143]: I think this warning is no longer needed?

File: locust_plugins/listeners.py [hunk 3]
```diff
@@ -130,22 +133,34 @@ def on_start(self, environment: locust.env.Environment):
...
         else:
             # non-swarm runs need to generate the run id here
             self._run_id = datetime.now(timezone.utc)
-        if not is_worker():
+            logging.info(f"Run id is {self._run_id}")
+        if not is_worker(self.env):
             logging.info(
                 f"Follow test run here: {self.env.parsed_options.grafana_url}&var-testplan={self._testplan}&from={int(self._run_id.timestamp()*1000)}&to=now"
             )
+            if type(environment.runner) != LocalRunner:
```
Comment (cyberw) [inline locust_plugins/listeners.py:156]: I think LocalRunner can send messages to itself, so a lot of this logic may not be necessary. Try it out at least.

Comment (Recolada) [reply locust_plugins/listeners.py:156]: i had tried it, but get no client for localrunner. so added the condition.
Traceback (most recent call last):
  File "/home/recolada/Desktop/locust/.venv/lib/python3.10/site-packages/locust/event.py", line 40, in fire
    handler(**kwargs)
  File "/home/recolada/Desktop/locust/.venv/lib/python3.10/site-packages/locust_plugins/listeners.py", line 159, in on_start
    for _, worker in enumerate(environment.runner.clients):
AttributeError: 'LocalRunner' object has no attribute 'clients'

Comment (cyberw) [reply locust_plugins/listeners.py:156]: (no need for a for-loop or anything)

Comment (Recolada) [reply locust_plugins/listeners.py]: yes!! i changed it as per comment.

File: locust_plugins/listeners.py [hunk 4]
```diff
@@ -79,6 +77,13 @@ def __init__(
         events.spawning_complete.add_listener(self.spawning_complete)
         atexit.register(self.log_stop_test_run)
 
+        if is_worker(self.env):
```
Comment (cyberw) [inline locust_plugins/listeners.py:80]: Probably this if statement is not needed.

File: locust_plugins/listeners.py [hunk 5]
```diff
@@ -504,9 +506,9 @@ def user_error(self, user_instance, exception, tb, **kwargs):
             self.function(user_instance, exception, tb, **kwargs)
 
 
-def is_worker():
-    return "--worker" in sys.argv
+def is_worker(env: locust.env.Environment):
```
Comment (cyberw) [inline locust_plugins/listeners.py:509]: Probably best to just remove the method entirely, now that it is kind of trivial.

**Comment by cyberw** [comment]
Nice. I cant test it out right now, but maybe @DennisKrone ? or I can do it some time next week.

**Comment by cyberw** [comment]
Looks like your pylint issue may have been related to pylint version? And maybe the black reformat of the file was caused by not running from root? (thus ignoring the settings in pyproject.toml)

File: locust_plugins/listeners.py (hunk 5 above)
Comment (Recolada) [reply locust_plugins/listeners.py]: done

File: locust_plugins/listeners.py (hunk 4 above)
Comment (Recolada) [reply locust_plugins/listeners.py]: done

File: locust_plugins/listeners.py [hunk 6]
```diff
@@ -79,6 +77,13 @@ def __init__(
         events.spawning_complete.add_listener(self.spawning_complete)
         atexit.register(self.log_stop_test_run)
 
+        if self.env.parsed_options.worker:
+            self.env.runner.register_message("get_run_id", self.set_run_id)
+
+    def set_run_id(self, environment, msg, **kwargs):
+        logging.info(f'Received run id from master. Using this {datetime.strptime(msg.data, "%Y-%m-%d, %H:%M:%S.%f")}')
```
Comment (cyberw) [inline locust_plugins/listeners.py:84]: I prefer debug level here

File: locust_plugins/listeners.py [hunk 7]
```diff
@@ -130,22 +135,36 @@ def on_start(self, environment: locust.env.Environment):
...
             if getattr(environment.parsed_options, "run_id", False):
                 self._run_id = parser.parse(environment.parsed_options.run_id)
+            elif hasattr(self, "_run_id"):
+                pass
             else:
                 logging.info(
                     "You are running distributed, but without swarm. run_id:s in Timescale will not match exactly between load gens"
                 )
                 self._run_id = datetime.now(timezone.utc)
+                logging.info(f"Run id is {self._run_id}")
```
Comment (cyberw) [inline locust_plugins/listeners.py:153]: Shortly after this, the url will be logged anyway. Maybe make this debug level?

File: locust_plugins/listeners.py [hunk 8]
```diff
@@ -130,22 +135,36 @@ def on_start(self, environment: locust.env.Environment):
...
         else:
             # non-swarm runs need to generate the run id here
             self._run_id = datetime.now(timezone.utc)
-        if not is_worker():
```
Comment (cyberw) [inline locust_plugins/listeners.py:145]: The comment above about non-swarm runs is not relevant any more, is it?

File: locust_plugins/listeners.py (hunk 6 above)
Comment (Recolada) [reply locust_plugins/listeners.py]: Updated.

File: locust_plugins/listeners.py (hunk 7 above)
Comment (Recolada) [reply locust_plugins/listeners.py:153]: updated.

File: locust_plugins/listeners.py (hunk 8 above)
Comment (Recolada) [reply locust_plugins/listeners.py:145]: yes, i think too. removed it as well. Fixed these 3 comments and pushed to pr.

**Comment by cyberw** [comment]
worked like a charm

**Pull request merged**